tracer/src/Datadog.Trace/Telemetry/TelemetryFactory.cs

```diff
@@ -154,20 +153,23 @@ public ITelemetryController CreateTelemetryController(ImmutableTracerSettings tr
             TelemetrySettings settings)
         {
             var transportManager = new TelemetryTransportManagerV2(telemetryTransports);
+            // The telemetry controller must be a singleton, so we initialize once
+            // Note that any dependencies initialized inside the controller are also singletons (by design)
             // Initialized once so if we create a new controller from this factory we get the same collector instances
-            var integrations = LazyInitializer.EnsureInitialized(ref _integrations)!;
-            var products = LazyInitializer.EnsureInitialized(ref _products)!;
-            var application = LazyInitializer.EnsureInitialized(ref _application)!;
-
-            return new TelemetryControllerV2(
-                Config,
-                _dependencies!,
-                integrations,
-                Metrics,
-                products,
-                application,
-                transportManager,
-                settings.HeartbeatInterval);
+            var controller = LazyInitializer.EnsureInitialized(
```

Comment:
Because the TelemetryControllerV2 immediately starts, you can't do that. `LazyInitializer.EnsureInitialized` guarantees that all threads see the same instance, but multiple instances could be created. You can use the overload that takes a synchronization object, or use a `Lazy<T>` instead.

Reply:
Fixed with a lock (to avoid the closure allocation) in https://github.com/DataDog/dd-trace-dotnet/pull/4354/commits/b0ef98d512b0248d3b681d0b07f25b4df5ac2f23